### ISSUE TITLE
Service monitor - color monitor

### ISSUE THREAD

**Issue by decknine** [jurialmunkey/plugin.video.themoviedb.helper#298]
RE: Kodi Matrix - Aeon nox silvo skin

I am having difficulty getting color monitor to work correctly. through reading the documentation i have used the below code to activate color monitor in "startup.xml":

```
<onload>Skin.ToggleSetting(TMDbHelper.EnableColors)</onload>
<onload>SetProperty(TMDbHelper.Colors.SourceImage,poster,Home)</onload>
```

and then defined where i want this to come into effect through "Includes_Backgrounds.xml":

```
<control type="image"> ###FLOOR PANEL COLOR###
<width>100%</width>
<height>1266</height>
<texture colordiffuse="$INFO[Window(Home).Property(TMDbHelper.ListItem.Colors.Main)]">floor.png</texture>
<visible>Container.Content(episodes) + !String.IsEmpty(ListItem.Title) | Container.Content(Seasons) | Container.Content(Tvshows) | Container.Content(Movies)</visible>
</control>
```

this worked once and now i cant seem to get it to work again...

**Comment by jurialmunkey** [comment]
Don't toggle the setting... Each time you startup you're flipping the switch on/off.

Use `Skin.SetBool(TMDbHelper.EnableColors)` instead.

You also need to set TMDbhelper.WidgetContainer property to the ID of the list you want to monitor if for a widget. And you need to make sure it is cleared in the library (otherwise it will be looking at whatever ID you set instead of the standard container)

**Comment by jurialmunkey** [comment]
Also note that PIL (the python library used for colour and blur) doesn't work on all systems in Matrix currently (should work on windows and maybe Linux based systems depending, but won't work on android)

**Comment by decknine** [comment]
i figured as much for the toggle... thanks for clearing that up! im not using a widget and i am running windows. using the code below i was able to get it working again:
`<onload condition="System.HasAddon(plugin.video.themoviedb.helper)">Skin.SetBool(TMDbHelper.Service)</onload>
	<onload condition="System.HasAddon(plugin.video.themoviedb.helper)">Skin.SetBool(TMDbHelper.EnableColors)</onload>
	<onload condition="System.HasAddon(plugin.video.themoviedb.helper)">SetProperty(TMDbHelper.Colors.SourceImage,poster,Home)</onload>
	
	<onload>Skin.SetString(TMDbHelper.Colors.Luminance,0.65)</onload>
	<onload>Skin.SetString(TMDbHelper.Colors.Saturation,0.55)</onload>`
without "Skin.SetString(TMDbHelper.Colors.Saturation,0.**)"  and luminance included, i wasn't able to get it to work at all. also interestingly if you set both values to "1.0" you only get the colour white... i assumed this was because of the maxed luminance but any setting above 0.75 results in white for me.

happy to get it working, thanks for the tips! 

is there any way to play around with different color results? e.g. picking the second most prominent color could be really cool.

![Capture](https://user-images.githubusercontent.com/42182450/93010238-e4d45980-f5bc-11ea-9233-f3e35ad11415.JPG)

**Comment by jurialmunkey** [comment]
The colour isn't the most "prominent" colour - it's the average of all pixel values (after resizing to a 100px square thumb).

Because of all the slight variations in values, if you actually got the most prominent colour it would nearly always end up as black or white - that's why need to average instead (plus quicker to average)

There is the comp colour though, which is a hue shifted version of main colour. 

I'll have to double check the lum/sat code because it should work even if the values aren't set. 

Also yeah high lum values will wash out colours. Have a look at this image to see how hsl works https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSrY3t9kUPx0vxhPi3agPB9qOQ7z9CRpWxvXA&usqp=CAU

**Comment by decknine** [comment]
Ok I see, it makes sense anything more advanced would be too resource
intensive in real time.

Maybe being able to set the comp reference shift amount manually could be
useful. Neighbouring colours etc.

Yeah I found it strange that the lum/sat amount wouldn't work at a value of
0.85

The code originally worked without defining those values but after updating
it today, with your changes, it wouldn't work without them... I'm sure user
error 😅

**Comment by jurialmunkey** [comment]
Latest version gives the option to set `TMDBHelper.Colors.CompShift` to shift the hue of comp colour. Value is between 0.0 and 1.0 with default for comp being 0.33 (120 degrees - triadic colour). Generally will only shift comp colour but sometimes might be useful also to shift main colour (e.g. for split complimentary colours), so I've also added in the option to set `TMDBHelper.Colors.MainShift` just in case.

Values other than default 0.33 that also make sense
Complimentary 0.5 (180 degrees)
Split complimentary 0.4167 and 0.5833 (150 and 210 degrees)
Analogous 0.25 (90 degrees) -- or use 0.75 for counterclockwise 
Close Analogous  0.167 (60 degrees) -- or use 0.833 for counterclockwise



Also I tested with luminance 0.85 and it appeared to be working fine for me with discernible difference in colours (though very washed out). For instance, the border on the below two posters was made with sat=0.5 and lum=0.85

![screenshot263](https://user-images.githubusercontent.com/4618719/93018736-37416480-f615-11ea-9eca-00ae10b49d1d.png)
![screenshot265](https://user-images.githubusercontent.com/4618719/93018737-390b2800-f615-11ea-9a0e-9951dd4c1a0d.png)
